test/Npgsql.Tests/SystemTransactionTests.cs

```diff
@@ -130,21 +162,106 @@ public void TwoConnections()
+                        "Unexpected first insert rowcount");
+                }
+
+                using (var conn2 = OpenConnection(ConnectionStringEnlistOn))
+                {
+                    Assert.That(conn2.ExecuteNonQuery(@"INSERT INTO data (name) VALUES ('test2')"), Is.EqualTo(1),
+                        "Unexpected second insert rowcount");
+                }
+
+                // Consecutive connections used in same scope should not promote the
+                // transaction to distributed.
+                AssertNoPreparedTransactions();
+                scope.Complete();
+            }
+            AssertNumberOfRows(2);
+        }
+
+        [Test]
         public void TwoConnectionsRollback()
         {
             using (new TransactionScope())
             using (var conn1 = OpenConnection(ConnectionStringEnlistOn))
             using (var conn2 = OpenConnection(ConnectionStringEnlistOn))
             {
-                Assert.That(conn1.ExecuteNonQuery(@"INSERT INTO data (name) VALUES ('test1')"), Is.EqualTo(1));
-                Assert.That(conn2.ExecuteNonQuery(@"INSERT INTO data (name) VALUES ('test2')"), Is.EqualTo(1));
+                Assert.That(conn1.ExecuteNonQuery(@"INSERT INTO data (name) VALUES ('test1')"), Is.EqualTo(1),
+                    "Unexpected first insert rowcount");
+                Assert.That(conn2.ExecuteNonQuery(@"INSERT INTO data (name) VALUES ('test2')"), Is.EqualTo(1),
+                    "Unexpected second insert rowcount");
             }
             // TODO: There may be a race condition here, where the prepared transaction above still hasn't committed.
             AssertNoPreparedTransactions();
             AssertNumberOfRows(0);
         }
 
         [Test]
+        public void DistributedRollback()
+        {
+            var disposedCalled = false;
+            var tx = new TransactionScope();
+            try
+            {
+                using (var conn1 = OpenConnection(ConnectionStringEnlistOn))
+                {
+                    Assert.That(conn1.ExecuteNonQuery(@"INSERT INTO data (name) VALUES ('test1')"), Is.EqualTo(1),
+                        "Unexpected first insert rowcount");
+
+                    ForceEscalationToDistributedTx.Escalate(true);
+                    tx.Complete();
+                }
+                disposedCalled = true;
+                Assert.That(() => tx.Dispose(), Throws.TypeOf<TransactionAbortedException>());
+                // TODO: There may be a race condition here, where the prepared transaction above still hasn't completed.
+                AssertNoPreparedTransactions();
+                AssertNumberOfRows(0);
+            }
+            finally
+            {
+                if (!disposedCalled)
+                    tx.Dispose();
+            }
+        }
+
+        [Test, Explicit("Failing test and 100 iteration loop.")]
+        public void DistributedRace()
```

Comment:
Here is the test demonstrating the race condition. I was not able of demonstrating it with the simpler `TwoConnections` test. It looks like enlisting with a durable ressource is required for triggering it.